### ISSUE TITLE
gcc-10 static analysis at 1e280ca517976f464116a460729cd0a34b5dd682

### ISSUE THREAD

**Issue by gdevenyi** [zlib-ng/zlib-ng#592]
```
$ gcc --version
gcc-10 (Ubuntu 10.1.0-2ubuntu1~18.04) 10.1.0
$ CC=gcc-10 cmake -DCMAKE_C_FLAGS_RELEASE:STRING='-O3 -DNDEBUG -fanalyzer -Wanalyzer-double-fclose -Wanalyzer-double-free -Wanalyzer-exposure-through-output-file -Wanalyzer-file-leak -Wanalyzer-free-of-non-heap -Wanalyzer-malloc-leak -Wanalyzer-null-argument -Wanalyzer-null-dereference -Wanalyzer-possible-null-argument -Wanalyzer-possible-null-dereference -Wanalyzer-stale-setjmp-buffer -Wanalyzer-tainted-array-index -Wanalyzer-unsafe-call-within-signal-handler -Wanalyzer-use-after-free -Wanalyzer-use-of-pointer-in-stale-stack-frame' .
make -j 40 2> gcc-analyzer.log
```

output:
```
In file included from /home/gdevenyi/scratch/zlib-ng/arch/x86/deflate_quick.c:29:
/home/gdevenyi/scratch/zlib-ng/arch/x86/../../trees_emit.h:75:13: warning: ‘bi_flush’ defined but not used [-Wunused-function]
   75 | static void bi_flush(deflate_state *s) {
      |             ^~~~~~~~
In file included from /home/gdevenyi/scratch/zlib-ng/arch/x86/deflate_quick.c:29:
/home/gdevenyi/scratch/zlib-ng/arch/x86/../../trees_emit.h:75:13: warning: ‘bi_flush’ defined but not used [-Wunused-function]
   75 | static void bi_flush(deflate_state *s) {
      |             ^~~~~~~~
/home/gdevenyi/scratch/zlib-ng/test/example.c: In function ‘test_gzio’:
/home/gdevenyi/scratch/zlib-ng/test/example.c:199:16: warning: pointer targets in passing argument 1 of ‘strcmp’ differ in signedness [-Wpointer-sign]
  199 |     if (strcmp(uncompr, hello) != 0) {
      |                ^~~~~~~
      |                |
      |                unsigned char *
In file included from /home/gdevenyi/scratch/zlib-ng/zutil.h:22,
                 from /home/gdevenyi/scratch/zlib-ng/deflate.h:13,
                 from /home/gdevenyi/scratch/zlib-ng/test/example.c:12:
/usr/include/string.h:136:32: note: expected ‘const char *’ but argument is of type ‘unsigned char *’
  136 | extern int strcmp (const char *__s1, const char *__s2)
      |                    ~~~~~~~~~~~~^~~~
/home/gdevenyi/scratch/zlib-ng/test/example.c:84:14: warning: variable ‘read’ set but not used [-Wunused-but-set-variable]
   84 |     int err, read;
      |              ^~~~
/home/gdevenyi/scratch/zlib-ng/test/example.c: In function ‘test_deflate_set_header’:
/home/gdevenyi/scratch/zlib-ng/test/example.c:867:16: warning: dereference of possibly-NULL ‘head’ [CWE-690] [-Wanalyzer-possible-null-dereference]
  867 |     head->text = 1;
      |     ~~~~~~~~~~~^~~
  ‘test_deflate_set_header’: event 1
    |
    |  853 |     PREFIX(gz_header) *head = calloc(256, 1);
    |      |                               ^~~~~~~~~~~~~~
    |      |                               |
    |      |                               (1) this call could return NULL
    |
  ‘test_deflate_set_header’: event 2
    |
    |   23 |     if (err != Z_OK) { \
    |      |        ^
    |      |        |
    |      |        (2) following ‘false’ branch (when ‘err == 0’)...
/home/gdevenyi/scratch/zlib-ng/test/example.c:865:5: note: in expansion of macro ‘CHECK_ERR’
    |  865 |     CHECK_ERR(err, "deflateInit2");
    |      |     ^~~~~~~~~
    |
  ‘test_deflate_set_header’: events 3-4
    |
    |  867 |     head->text = 1;
    |      |     ~~~~~~~~~~~^~~
    |      |                |
    |      |                (3) ...to here
    |      |                (4) ‘head’ could be NULL: unchecked value from (1)
    |
/home/gdevenyi/scratch/zlib-ng/test/example.c: In function ‘test_gzio’:
/home/gdevenyi/scratch/zlib-ng/test/example.c:199:16: warning: pointer targets in passing argument 1 of ‘strcmp’ differ in signedness [-Wpointer-sign]
  199 |     if (strcmp(uncompr, hello) != 0) {
      |                ^~~~~~~
      |                |
      |                unsigned char *
In file included from /home/gdevenyi/scratch/zlib-ng/zutil.h:22,
                 from /home/gdevenyi/scratch/zlib-ng/deflate.h:13,
                 from /home/gdevenyi/scratch/zlib-ng/test/example.c:12:
/usr/include/string.h:136:32: note: expected ‘const char *’ but argument is of type ‘unsigned char *’
  136 | extern int strcmp (const char *__s1, const char *__s2)
      |                    ~~~~~~~~~~~~^~~~
/home/gdevenyi/scratch/zlib-ng/test/example.c:84:14: warning: variable ‘read’ set but not used [-Wunused-but-set-variable]
   84 |     int err, read;
      |              ^~~~
/home/gdevenyi/scratch/zlib-ng/test/example.c: In function ‘test_deflate_set_header’:
/home/gdevenyi/scratch/zlib-ng/test/example.c:867:16: warning: dereference of possibly-NULL ‘head’ [CWE-690] [-Wanalyzer-possible-null-dereference]
  867 |     head->text = 1;
      |     ~~~~~~~~~~~^~~
  ‘test_deflate_set_header’: event 1
    |
    |  853 |     PREFIX(gz_header) *head = calloc(256, 1);
    |      |                               ^~~~~~~~~~~~~~
    |      |                               |
    |      |                               (1) this call could return NULL
    |
  ‘test_deflate_set_header’: event 2
    |
    |   23 |     if (err != Z_OK) { \
    |      |        ^
    |      |        |
    |      |        (2) following ‘false’ branch (when ‘err == 0’)...
/home/gdevenyi/scratch/zlib-ng/test/example.c:865:5: note: in expansion of macro ‘CHECK_ERR’
    |  865 |     CHECK_ERR(err, "deflateInit2");
    |      |     ^~~~~~~~~
    |
  ‘test_deflate_set_header’: events 3-4
    |
    |  867 |     head->text = 1;
    |      |     ~~~~~~~~~~~^~~
    |      |                |
    |      |                (3) ...to here
    |      |                (4) ‘head’ could be NULL: unchecked value from (1)
    |

```

**Comment by nmoinvaz** [comment]
Some of these are fixed in #582

**Comment by gdevenyi** [comment]
Here's -fanalyzer after #582 
```
In file included from /home/gdevenyi/scratch/zlib-ng/arch/x86/deflate_quick.c:29:
/home/gdevenyi/scratch/zlib-ng/arch/x86/../../trees_emit.h:75:13: warning: ‘bi_flush’ defined but not used [-Wunused-function]
   75 | static void bi_flush(deflate_state *s) {
      |             ^~~~~~~~
In file included from /home/gdevenyi/scratch/zlib-ng/arch/x86/deflate_quick.c:29:
/home/gdevenyi/scratch/zlib-ng/arch/x86/../../trees_emit.h:75:13: warning: ‘bi_flush’ defined but not used [-Wunused-function]
   75 | static void bi_flush(deflate_state *s) {
      |             ^~~~~~~~
/home/gdevenyi/scratch/zlib-ng/test/example.c: In function ‘test_gzio’:
/home/gdevenyi/scratch/zlib-ng/test/example.c:84:14: warning: variable ‘read’ set but not used [-Wunused-but-set-variable]
   84 |     int err, read;
      |              ^~~~
/home/gdevenyi/scratch/zlib-ng/test/example.c: In function ‘test_deflate_set_header’:
/home/gdevenyi/scratch/zlib-ng/test/example.c:872:16: warning: dereference of possibly-NULL ‘head’ [CWE-690] [-Wanalyzer-possible-null-dereference]
  872 |     head->text = 1;
      |     ~~~~~~~~~~~^~~
  ‘test_deflate_set_header’: event 1
    |
    |  858 |     PREFIX(gz_header) *head = calloc(256, 1);
    |      |                               ^~~~~~~~~~~~~~
    |      |                               |
    |      |                               (1) this call could return NULL
    |
  ‘test_deflate_set_header’: event 2
    |
    |   23 |     if (err != Z_OK) { \
    |      |        ^
    |      |        |
    |      |        (2) following ‘false’ branch (when ‘err == 0’)...
/home/gdevenyi/scratch/zlib-ng/test/example.c:870:5: note: in expansion of macro ‘CHECK_ERR’
    |  870 |     CHECK_ERR(err, "deflateInit2");
    |      |     ^~~~~~~~~
    |
  ‘test_deflate_set_header’: events 3-4
    |
    |  872 |     head->text = 1;
    |      |     ~~~~~~~~~~~^~~
    |      |                |
    |      |                (3) ...to here
    |      |                (4) ‘head’ could be NULL: unchecked value from (1)
    |
/home/gdevenyi/scratch/zlib-ng/test/example.c: In function ‘test_gzio’:
/home/gdevenyi/scratch/zlib-ng/test/example.c:84:14: warning: variable ‘read’ set but not used [-Wunused-but-set-variable]
   84 |     int err, read;
      |              ^~~~
/home/gdevenyi/scratch/zlib-ng/test/example.c: In function ‘test_deflate_set_header’:
/home/gdevenyi/scratch/zlib-ng/test/example.c:872:16: warning: dereference of possibly-NULL ‘head’ [CWE-690] [-Wanalyzer-possible-null-dereference]
  872 |     head->text = 1;
      |     ~~~~~~~~~~~^~~
  ‘test_deflate_set_header’: event 1
    |
    |  858 |     PREFIX(gz_header) *head = calloc(256, 1);
    |      |                               ^~~~~~~~~~~~~~
    |      |                               |
    |      |                               (1) this call could return NULL
    |
  ‘test_deflate_set_header’: event 2
    |
    |   23 |     if (err != Z_OK) { \
    |      |        ^
    |      |        |
    |      |        (2) following ‘false’ branch (when ‘err == 0’)...
/home/gdevenyi/scratch/zlib-ng/test/example.c:870:5: note: in expansion of macro ‘CHECK_ERR’
    |  870 |     CHECK_ERR(err, "deflateInit2");
    |      |     ^~~~~~~~~
    |
  ‘test_deflate_set_header’: events 3-4
    |
    |  872 |     head->text = 1;
    |      |     ~~~~~~~~~~~^~~
    |      |                |
    |      |                (3) ...to here
    |      |                (4) ‘head’ could be NULL: unchecked value from (1)
    |

```
